src/package.ts

```diff
@@ -282,69 +312,55 @@ export class Package extends AsyncOptionalCreatable {
         }
         const version = dependencies[name];
-
-        if (version.startsWith('npm:')) {
-          return {
-            name: parseAliasedPackageName(version),
-            version: version.split('@').reverse()[0].replace('^', '').replace('~', ''),
-            alias: name,
-            tag: tag || targetTag,
-          };
-        } else {
-          return {
-            name,
-            version: version.split('@').reverse()[0].replace('^', '').replace('~', ''),
-            alias: null,
-            tag: tag || targetTag,
-          };
-        }
+        return getPinnedPackage({ name, version, tag, targetTag });
       })
-      .filter((d) => !!d);
+      .filter(isObject);
 
-    const pinnedPackages: PinnedPackage[] = [];
-    deps.forEach((dep) => {
-      // get the 'release' tag version or the version specified by the passed in tag
-      const versions = this.getDistTags(dep.name);
-      let tag = dep.tag;
+    const updatedDeps = this.calculatePinnedPackageUpdates(deps);
 
-      // if tag is 'latest-rc' and there's no latest-rc release for a package, default to latest
-      if (!versions[tag]) {
-        tag = 'latest';
-      }
-
-      // If the version in package.json is greater than the version of the requested tag, then we
-      // assume that this is on purpose - so we don't overwrite it. For example, we might want to
-      // include a latest-rc version for a single plugin but everything else we want latest.
-      let version: string;
-      if (semver.gt(dep.version, versions[tag])) {
-        CliUx.ux.warn(
-          `${dep.name} is currently pinned at ${dep.version} which is higher than ${tag} (${versions[tag]}). Assuming that this is intentional...`
-        );
-        version = dep.version;
-        tag = findKey(versions, (v) => v === version);
+    updatedDeps.forEach((pp) => {
+      if (pp.alias) {
+        this.packageJson.dependencies[pp.alias] = `npm:${pp.name}@${pp.version}`;
       } else {
-        version = versions[tag];
+        this.packageJson.dependencies[pp.name] = pp.version;
       }
+    });
+    return updatedDeps;
+  }
 
-      // insert the new hardcoded versions into the dependencies in the project's package.json
-      if (dep.alias) {
-        this.packageJson.dependencies[dep.alias] = `npm:${dep.name}@${version}`;
+  public bumpJit(targetTag = 'latest-rc'): ChangedPackageVersions {
+    // no JIT is ok
+    if (!this.packageJson.oclif.jitPlugins) {
+      return;
+    }
+
+    const { pinnedDependencies, dependencies, devDependencies } = this.packageJson;
+    const { jitPlugins } = this.packageJson.oclif;
+    const jitDeps = Object.entries(jitPlugins)
+      .map(([plugin, version]) => {
+        const [name, tag] = getNameAndTag(plugin);
+        if (dependencies?.[name] || devDependencies?.[name] || pinnedDependencies?.includes(name)) {
+          throw new SfError('JIT plugins should not be listed in dependencies, devDependencies, or pinnedDependencies');
```

Comment:
What about `oclif.plugins`? Do we need to make sure that the jit plugin does not exist there also?

Reply:
good idea.